src/objects/npc.lua

```diff
@@ -0,0 +1,175 @@
+        end
+
+        if self.moveintent == 1 then
+            self.dx = self.dx + self.dx_accel * dt
+            self.is_walking = true
+            self.direction = 'right'
+            self.spr = self.walk
+        end
+
+        -- perform jumping if we can/should
+        if self.jumpintent == 0 and self.jump_enabled then
+            self.dy = self.jump_dy
+            self.jump_enabled = false
+            self.jumpintent = 1
+        end
+
+        -- slow down
+            if not self.jump_enabled and not self.is_walking then
+                if self.dx > 0 then
+                    self.dx = math.max(self.dx - (self.passive_decel / 1.5) * dt, 0)
+                else
+                    self.dx = math.min(self.dx + (self.passive_decel / 1.5) * dt, 0)
+                end
+            else
+                if self.jump_enabled and not self.is_walking then
+                    if self.dx > 0 then
+                        self.dx = math.max(self.dx - self.passive_decel * dt, 0)
+                    else
+                        self.dx = math.min(self.dx + self.passive_decel * dt, 0)
+                    end
+                end
+            end
+
+        -- limit maximum horizontal speed
+        if self.dx > self.dx_max then
+            self.dx = self.dx_max
+        elseif self.dx < -self.dx_max then
+            self.dx = -self.dx_max
+        end
+
+        -- apply gravity
+        self.dy = self.dy + self.gravity * dt
+
+        -- if moving suspiciously vertically then disable jumping
+        if math.abs(self.dy) > 20 then
+            self.jump_enabled = false
+        end
+
+        -- resolve new velocity
+        -- first, resolve horizontal movement
+
+        self.x = self.x + self.dx * dt
+        local collision = obj.get_collisions(self, self.geometry_layer, true)
+
+        if collision ~= nil then
+            -- resolve a horizontal collision, depending on direction of movement
+            if self.dx > 0 then
+                self.x = collision.x - (self.w + 1)
+            elseif self.dx < 0 then
+                self.x = collision.x + collision.w + 1
+            else
+                assert(false) -- if this ever happens then there are problems
+            end
+
+            self.dx = 0
+        end
+
+        -- now, resolve vertical movement
+
+        self.y = self.y + self.dy * dt
+        collision = obj.get_collisions(self, self.geometry_layer, true)
+
+        if collision ~= nil then
+            if self.dy >= 0 then
+                -- here it's safer to assume dy==0 -> the npc was moving somewhat down-ish
+                self.y = collision.y - (self.h + 1)
+                self.jump_enabled = true
+            else
+                self.y = collision.y + collision.h + 1
+            end
+
+            self.dy = 0
+        end
+
+    end,
+    render = function(self)
+
+        -- clamp rendering to integers, otherwise fuzzy collisions
+        -- end up making the sprite look all jittery
+
```

Comment:
It would be good to set the color to white here before drawing the sprite to prevent interference (when the player turns yellow, the NPC will as well)

Reply:
I'm actually doing something similar here already.